swap/protocol.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	contract "github.com/ethersphere/swarm/contracts/swap"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+// ErrEmptyAddressInSignature is used when the empty address is used for the chequebook in the handshake
+var ErrEmptyAddressInSignature = errors.New("empty address in handshake")
+
+// ErrInvalidHandshakeMsg is used when the message received during handshake does not conform to the
+// structure of the HandshakeMsg
+var ErrInvalidHandshakeMsg = errors.New("invalid handshake message")
+
+// Spec is the swap protocol specification
+var Spec = &protocols.Spec{
+	Name:       "swap",
+	Version:    1,
+	MaxMsgSize: 10 * 1024 * 1024,
+	Messages: []interface{}{
+		HandshakeMsg{},
+		EmitChequeMsg{},
+	},
+}
+
+// Protocols is a node.Service interface method
+func (s *Swap) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{
+		{
+			Name:    Spec.Name,
+			Version: Spec.Version,
+			Length:  Spec.Length(),
+			Run:     s.run,
+		},
+	}
+}
+
+// APIs is a node.Service interface method
+func (s *Swap) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "swap",
+			Version:   "1.0",
+			Service:   NewAPI(s),
+			Public:    false,
+		},
+	}
+}
+
+// Start is a node.Service interface method
+func (s *Swap) Start(server *p2p.Server) error {
+	log.Info("Swap service started")
+	return nil
+}
+
+// Stop is a node.Service interface method
+func (s *Swap) Stop() error {
+	return nil
+}
+
+// verifyHandshake verifies the chequebook address transmitted in the swap handshake
+func (s *Swap) verifyHandshake(msg interface{}) error {
+	handshake, ok := msg.(*HandshakeMsg)
+	if !ok {
+		return ErrInvalidHandshakeMsg
+	}
+
+	if (handshake.ContractAddress == common.Address{}) {
+		return ErrEmptyAddressInSignature
+	}
+
+	return s.verifyContract(context.TODO(), handshake.ContractAddress)
```

Comment:
`context.Background()` (unless you have some other, different, future intentions)

Reply:
fixed